util/systemd/systemd_linux.go

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build linux
+
+package systemd
+
+import (
+	"log"
+	"sync"
+
+	"github.com/mdlayher/sdnotify"
+)
+
+var getNotifyOnce struct {
+	sync.Once
+	v *sdnotify.Notifier
+}
+
+type logOnce struct {
+	sync.Once
+}
+
+func (l *logOnce) logf(format string, args ...interface{}) {
+	l.Once.Do(func() {
+		log.Printf(format, args...)
+	})
+}
+
+var (
+	readyOnce  = &logOnce{}
+	statusOnce = &logOnce{}
+)
+
+func notifier() *sdnotify.Notifier {
+	getNotifyOnce.Do(func() {
+		var err error
+		getNotifyOnce.v, err = sdnotify.New()
+		if err != nil {
+			log.Printf("systemd: systemd-notifier error: %v", err)
+		}
+	})
+	return getNotifyOnce.v
+}
+
+// Ready signals readiness to systemd. This will unblock service dependents from starting.
+func Ready() {
+	err := notifier().Notify(sdnotify.Ready)
```

Comment:
I think this'll panic if sdnotify.New() returned nil above (unless the sdnotify package was explicitly designed to work correctly given nil receivers). I think you'll need a bit more if != nil boilerplate here and in Status.

Reply:
The sdnotify package was explicitly designed to work correctly given nil recievers: https://github.com/mdlayher/sdnotify/blob/master/sdnotify.go#L77-L79